tools/dochack/fuzzysearch.nim

```diff
@@ -0,0 +1,175 @@
+# A Fuzzy Match implementation inspired by the sublime text fuzzy match algorithm
+# as described here: https://blog.forrestthewoods.com/reverse-engineering-sublime-text-s-fuzzy-match-4cffeed33fdb
+# Heavily modified to provide more subjectively useful results
+# for on the Nim manual.
+#
+
+import strutils
+import math
+import macros
+
+type
+  FuzzyMatchResult* = object
+    ismatch*: bool
+    score*: int 
+
+
+# Fuzzy match score.
+#  A positive number is a bonus
+#  A negative number is a penalty
+type ScoreCard {.pure.} = enum 
+  Start = -100, # this acts as a state machine. So we need nil start and end states.
+  UnmatchedLeadingChar = -3, # for each unmatched character from the start of the string
+  UnmatchedChar = -1, # for each single letter does not match
+  MatchedChar = 0, # for each single letter matches
+  ConsecutiveMatch = 5, # every consecutive letter that matches after the first
+  LeadingCharMatch = 10, # The character matches the begining of the string 
+                         # or the first character of a word boundry (a space)
+                         # or a camel case boundry: 
+                         #    An upper case letter that is preceded by a lower case letter.
+                         # This bonus is only applied if a consecutive match is also found.
+                         # This is because it gives too much bias
+                         # to single char matches at word boundries.
+  WordBoundryMatch = 20
+
+const
+  MaxUnmatchedLeadingChar = 3
+  # max number of times we will apply the penalty for unmatched leading chars.
+
+  HeadingScaleFactor = 0.5
+  # The score from before the colon Char is multiplied by this.
+  # This is to weight function signatures and descriptions over module titles.
+
+template stateTransition(currentState, nextState: ScoreCard, score: int): untyped =
+  currentState = nextState
+  score += ord(nextState)
+
+proc fuzzymatch* (pattern, str: cstring) : FuzzyMatchResult =
+  var
+    scoreState = ScoreCard.Start
+    headerMatched = false
+    unmatchedLeadingCharCount = 0
+    consecutiveMatchCount = 0
+    strIndex = 0
+    patIndex = 0
+    curScore = 0
+    normalizedStrChar : char
+    normalizedPatternChar : char
+
+  while (strIndex < str.len) and (patIndex < pattern.len):
+    # normalize from strutils doesn't work on cstrings :-(
+    # I am doing an inline emulation of it here.
+    # lowercase everything and ignore underscores.
+    normalizedPatternChar = pattern[patIndex].toLowerAscii
+    normalizedStrChar = str[strIndex].toLowerAscii
+    #ignore certain chars
+    if
```

Comment:
Terrible indentation here.

Reply:
@Varriount helped me fix. The new version is much reworked (hopefully for the better).